libsodium-bindings/src/LibSodium/Bindings/SecretStream.hs

```diff
@@ -0,0 +1,267 @@
+-- * Messages can have different sizes.
+-- * There are no practical limits to the total length of the stream, or to the total number of individual messages.
+-- * Ratcheting: at any point in the stream, it is possible to "forget" the key used to encrypt the previous messages, and switch to a new key.
+
+-- $usage
+-- An encrypted stream starts with a short header, whose size is 'cryptoSecretStreamXChaCha20Poly1305HeaderBytes' bytes.
+-- That header must be sent/stored before the sequence of encrypted messages, as it is required to decrypt the stream.
+-- The header content doesn't have to be secret and decryption with a different header would fail.
+--
+-- A tag is attached to each message. That tag can be any of:
+--
+-- * 0, or 'cryptoSecretStreamXChaCha20Poly1305TagMessage': the most common tag, that doesn't add any information about the nature of the message.
+--
+-- * 'cryptoSecretStreamXChaCha20Poly1305TagFinal': indicates that the message marks the end of the stream, and erases the secret key used to encrypt the previous sequence.
+--
+-- * 'cryptoSecretStreamXChaCha20Poly1305TagPush': indicates that the message marks the end of a set of messages, but not the end of the stream.
+--   For example, a huge JSON string sent as multiple chunks can use this tag to indicate to the application that the string is complete and that it
+--   can be decoded. But the stream itself is not closed, and more data may follow.
+--
+-- * 'cryptoSecretStreamXChaCha20Poly1305TagRekey': "forget" the key used to encrypt this message and the previous ones, and derive a new secret key.
+--
+-- A typical encrypted stream simply attaches 0 as a tag to all messages, except the last one which is tagged as 'cryptoSecretStreamXChaCha20Poly1305TagFinal'.
+--
+-- Note that tags are encrypted; encrypted streams do not reveal any information about sequence boundaries
+-- ('cryptoSecretStreamXChaCha20Poly1305TagPush' and 'cryptoSecretStreamXChaCha20Poly1305TagRekey' tags).
+--
+-- For each message, additional data can be included in the computation of the authentication tag.
+-- With this API, additional data is rarely required, and most applications can just use 'Foreign.Ptr.nullPtr' and a length of 0 instead.
+
+-- | Opaque tag representing the hash state struct @crypto_secretstream_xchacha20poly1305_state@ used by the C API.
+--
+-- To use a 'CryptoSecretStreamXChaCha20Poly1305State', use 'withCryptoSecretStreamXChaCha20Poly1305State'.
+--
+-- @since 0.0.1.0
+data CryptoSecretStreamXChaCha20Poly1305State
+
+-- | Allocate an opaque 'CryptoSecretStreamXChaCha20Poly1305State' of size 'cryptoSecretStreamXChaCha20Poly1305StateBytes'.
+--
+-- @since 0.0.1.0
+withCryptoSecretStreamXChaCha20Poly1305State :: (Ptr CryptoSecretStreamXChaCha20Poly1305State -> IO a) -> IO a
+withCryptoSecretStreamXChaCha20Poly1305State action = allocaBytes (fromIntegral cryptoSecretStreamXChaCha20Poly1305StateBytes) action
```

Comment:
This is dangerous: if someone leads the `CryptoSecretStreamXChaCha20Poly1305State` to the world, they'll be aiming at deallocated memory. This should have a note saying 'please do not do that thanks'.